### ISSUE TITLE
NonbondedForce offsets are not moved when setting force group

### ISSUE THREAD

**Issue by zhang-ivy** [openmm/openmm#3310]
I'm seeing what I think is undesired behavior when I:
1) create a system with a nonbonded force with offsets,
2) move the nonbonded force to force group 31,
3) evaluate the energy at force group 0

The energy should be 0 for force group 0, but I'm seeing that the energy is -64719.72950761177. 

When I do not add the offsets, I see that the energy is 0, as expected.

It seems like the offsets are not being moved to force group 31?

Here's my code for creating the systems (which are attached as an xml):
```python
from openmmtools.testsystems import AlanineDipeptideExplicit
from simtk import unit, openmm
from openmmtools.constants import kB

#############################################
# CONSTANTS
#############################################
temperature = 300.0 * unit.kelvin
kT = kB * temperature
beta = 1.0/kT
DEFAULT_PLATFORM = openmm.Platform.getPlatformByName("CUDA")

# Grab reference system
atp = AlanineDipeptideExplicit()

# Create system from scratch
system = openmm.System()

# Copy particles from reference to new system
for particle_idx in range(atp.system.getNumParticles()):
    particle_mass = atp.system.getParticleMass(particle_idx)
    system.addParticle(particle_mass)

# Create nonbonded force
standard_nonbonded_force = openmm.NonbondedForce()
system.addForce(standard_nonbonded_force)

# Add global parameters
standard_nonbonded_force.addGlobalParameter("lambda_alchemical_electrostatics_reciprocal", 0.0)

# Set nonbonded method
old_nbf = atp.system.getForce(3)
standard_nonbonded_force.setNonbondedMethod(old_nbf.getNonbondedMethod())

# Copy particles from reference system to new system's nonbonded force
for particle_index in range(system.getNumParticles()):

    charge, sigma, epsilon = old_nbf.getParticleParameters(particle_index)

    # Add particle to the standard nonbonded force
    standard_nonbonded_force.addParticle(charge, sigma * 0.0, epsilon * 0.0)

    # Charge and epsilon will be turned on at lambda = 0 and turned off at lambda = 1
    standard_nonbonded_force.addParticleParameterOffset('lambda_alchemical_electrostatics_reciprocal', particle_index, -charge, 0.0, 0.0)

```

Here is the code I use to check the energy:
```python

def compute_potential_components(system, positions, beta=beta, platform=DEFAULT_PLATFORM):
    """
    Compute potential energy, raising an exception if it is not finite.
    Parameters
    ----------
    context : simtk.openmm.Context
        The context from which to extract, System, parameters, and positions.
    """
 
    from perses.dispersed.utils import configure_platform
    platform = configure_platform(platform.getName(), fallback_platform_name='Reference', precision='double')

    # Set force group to 31
    force = system.getForce(0)
    force.setForceGroup(31)

    # Create new Context.
    integrator = openmm.VerletIntegrator(1.0 * unit.femtoseconds)
    context = openmm.Context(system, integrator, platform)
    context.setPositions(positions)
    
    # Retrieve energy for force group 0
    groups = set([0])
    potential = beta * context.getState(getEnergy=True, groups=groups).getPotentialEnergy()
    del context, integrator
    
    print(potential)

compute_potential_components(system, atp.positions)
```
[offset_problems.zip](https://github.com/openmm/openmm/files/7456010/offset_problems.zip)

**Comment by zhang-ivy** [comment]
Tagging @peastman for feedback!

**Comment by peastman** [comment]
I can't reproduce this.  When I run the code above, it prints 0.0.  Are you sure that's exactly the code you're running?

**Comment by zhang-ivy** [comment]
Hmm, you're running the exact code I shared (including using CUDA)? 

I cannot reproduce the problem on Reference or OpenCL.

**Comment by zhang-ivy** [comment]
Here's the environment I'm using: [perses-rest-over-protocol.yml.zip](https://github.com/openmm/openmm/files/7470833/perses-rest-over-protocol.yml.zip)

**Comment by peastman** [comment]
Yes.  The only thing I changed was to replace

```python
    from perses.dispersed.utils import configure_platform
    platform = configure_platform(platform.getName(), fallback_platform_name='Reference', precision='double')
```

by

```python
    openmm.Platform.getPlatformByName('CUDA').setPropertyDefaultValue('Precision', 'double')
```

since I don't have Perses installed.

**Comment by zhang-ivy** [comment]
Hmm, I'm confused. I swapped out 
```python
from perses.dispersed.utils import configure_platform
platform = configure_platform(platform.getName(), fallback_platform_name='Reference', precision='double')
```

for 
```python
platform = openmm.Platform.getPlatformByName('CUDA')
platform.setPropertyDefaultValue('Precision', 'double')

```

and am still getting a nonzero energy. 

@ijpulidos : Would you mind trying to see if you can reproduce this problem (both on lilac and on your local GPU)?

**Comment by zhang-ivy** [comment]
I just create a fresh environment without perses and am still seeing the same problem:
```
mamba create -n debug-offsets -c conda-forge -c openeye openmm jupyter
mamba install -c conda-forge openmmtools # forgot to install openmmtools
```
[debug-offsets.yml.zip](https://github.com/openmm/openmm/files/7471010/debug-offsets.yml.zip)

**Comment by peastman** [comment]
I'm using the latest OpenMM code from the repository.  Can you try a current dev build?

**Comment by ijpulidos** [comment]
I can see it being non-zero with OpenMM 7.6, whereas using 7.5 does give `0.0`. I guess it's something that changed in 7.6 and fixed already in latest nightly build?

**Comment by ijpulidos** [comment]
FYI: Using latest openmm package from `omnia-dev` channel gives the same non-zero result.

@peastman Are these the latest nightly/dev builds?

**Comment by zhang-ivy** [comment]
Just created a fresh env with:
```
mamba create -n debug-offsets-dev -c omnia-dev/label/cuda102  -c conda-forge -c openeye openmm jupyter openmmtools
```
and I am still seeing the nonzero value.

**Comment by zhang-ivy** [comment]
Hmm.. it seems that the the `omnia-dev` packages haven't been updated in ~2 weeks...

**Comment by zhang-ivy** [comment]
Jaime did the following to [fix this in September](https://github.com/openmm/openmm/pull/3236#issuecomment-923719533) , but I don't know how to repeat his fix / don't have permissions

**Comment by zhang-ivy** [comment]
I just glanced at the PRs that have been merged in the last 16 days (which is when `omnia-dev` was last updated): 
![Screen Shot 2021-11-03 at 6 56 59 PM](https://user-images.githubusercontent.com/35546250/140225685-330c125a-c89d-4f31-ac89-9629a03445f1.png)

None of these seem like they would contain a change that would have fixed the issue I am raised though, so I'm still confused as to why @peastman cannot reproduce the issue

**Comment by peastman** [comment]
I take it the dev builds have stopped working again.  Sigh.

Could you try a couple of experiments?  First, try switching the nonbonded method to CutoffPeriodic.  Does that make the problem disappear?  This will tell us if it's something related to PME.  Second, try calling `setUseDispersionCorrection(False)` on the NonbondedForce.  That will tell us if it's something related to the dispersion correction.

**Comment by jchodera** [comment]
> I take it the dev builds have stopped working again. Sigh.

I'm on it: https://github.com/omnia-md/conda-dev-recipes/issues/208

> I'm using the latest OpenMM code from the repository. Can you try a current dev build?

@peastman : While I'm fixing the dev builds, so that we can establish whether or not it's a hardware-related issue, can you try the `conda` install and verify that you still cannot reduce the problem with released code on your hardware? =

https://github.com/openmm/openmm/issues/3310#issuecomment-960242054

**Comment by jchodera** [comment]
> I take it the dev builds have stopped working again. Sigh.

@peastman: OpenMM nightly dev builds are being pushed again now---will take another hour to get them all up:
https://anaconda.org/omnia-dev/openmm/files

**Comment by zhang-ivy** [comment]
> First, try switching the nonbonded method to CutoffPeriodic. Does that make the problem disappear? This will tell us if it's something related to PME. 
The problem goes away when I switch to CutoffPeriodic (energy is 0).

> Second, try calling setUseDispersionCorrection(False) on the NonbondedForce. That will tell us if it's something related to the dispersion correction.
The problem persists when I call setUseDispersionCorrection(False)

**Comment by peastman** [comment]
I still can't reproduce this.  Very strange!  Could you try a couple more experiments?

1. Try calling `getState()` twice.  Do the first and second calls report the same energy?
2. Try calling `setReciprocalSpaceForceGroup(1)` to put the reciprocal space calculation into its own group.  Does that affect the result?  What energies does it report for each of group 0, group 1, and group 31?

Thanks for fixing the build!

**Comment by zhang-ivy** [comment]
> I still can't reproduce this. Very strange! 

You tried installing from the nightly builds (after John fixed them) and you still can't reproduce?
I just created a fresh env using `mamba create -n debug-offset-dev2 -c omnia-dev/label/cuda102  -c conda-forge -c openeye openmm jupyter openmmtools` and am still able to reproduce the problem.

>Try calling getState() twice. Do the first and second calls report the same energy?

Yes they do

>Try calling setReciprocalSpaceForceGroup(1) to put the reciprocal space calculation into its own group. Does that affect the result? What energies does it report for each of group 0, group 1, and group 31?

force group 0: -64719.72950761177
force group 1: -64109.26402219434
force group 31: -327090.5169448105

**Comment by zhang-ivy** [comment]
@peastman : Are you able to reproduce the problem with `mamba create -n debug-offset-dev2 -c omnia-dev/label/cuda102 -c conda-forge -c openeye openmm jupyter openmmtools`?

**Comment by peastman** [comment]
I've been working on this, and it's very puzzling.  I can reproduce it if I install the conda build.  But if I compile myself from the same git revision, it works.  That makes it very hard to debug!

The energy being reported is the Ewald self energy.  Usually that's just a constant that's precomputed.  But when using offset parameters it can change between steps, so instead it gets computed on the GPU.  For that reason, it's not surprising for offsets to make a difference in behavior.  But when the reciprocal space force group isn't being computed, that energy isn't supposed to be added.  And when I compile the code myself, everything works exactly as it's supposed to.  I don't understand why the conda build works differently.

**Comment by jchodera** [comment]
> And when I compile the code myself, everything works exactly as it's supposed to. I don't understand why the conda build works differently.

Does this suggest some sort of memory initialization issue, where one compiler initialized memory to zeros and another one just doesn't initialize?

**Comment by peastman** [comment]
It seems unlikely.  Anyway, I can't figure out any way that would cause it.  We decide whether to include the self energy based only on two arguments that are passed to the function.

https://github.com/openmm/openmm/blob/0993868a93bb8a3db36b42da60adf4a846dbd3e4/platforms/cuda/src/CudaKernels.cpp#L1149-L1153

`includeReciprocal` ought to be false.  In that case, the kernel doesn't record the energy.

https://github.com/openmm/openmm/blob/0993868a93bb8a3db36b42da60adf4a846dbd3e4/platforms/common/src/kernels/nonbondedParameters.cc#L63-L64

Could it somehow be related to the definitions of the types?  The CPU code defines `computeSelfEnergy` as a bool while the GPU kernel defines it as an int.  But that seems improbable too.

**Comment by peastman** [comment]
I merged #3330, which makes the argument type consistent.  I'm hoping that will fix the problem.  If you can try tomorrow's dev build, we'll find out.

**Comment by zhang-ivy** [comment]
I just tested the nightly build, and am no longer seeing the problem! The energy of force group 0 is now 0, which is what I would expect.

**Comment by peastman** [comment]
Yay!  I think that means we can close this!

**Comment by jchodera** [comment]
We might be running into this issue impacting YANK (https://github.com/choderalab/yank/issues/1261) since this issue appears to have been introduced in the OpenMM 7.6 release. This means that we should note that this is a "Known Issue" with 7.6 in the release notes, and that we might consider back-porting a bugfix to 7.6.1 that rectifies this.

**Comment by peastman** [comment]
Why would we create a 7.6.1 release when we're just about to release 7.7?

**Comment by jchodera** [comment]
Is our policy that all users should abandon all previous major.minor releases as soon as a new release comes out?

**Comment by peastman** [comment]
Correct.  We only fix bugs in the latest release.  We don't provide any kind of extended support for older releases.

**Comment by peastman** [comment]
> Where do we keep the list of known bugs with OpenMM 7.6, then?

This issue tracker.

**Comment by jchodera** [comment]
We closed the issue because the bug has been addressed in a development branch, but there are folks using OpenMM 7.6 to generate problematic results right now (e.g. https://github.com/choderalab/yank/issues/1261). We should presumably add "Known bugs" to the [OpenMM 7.6 release notes](https://github.com/openmm/openmm/releases/tag/7.6.0) because we have elected not to fix these, and let people know that this is a known issue that is specific to OpenMM 7.6.
